Builder wrapper exit code 1 masks MCP failure from Python retry logic

## Problem

When `claude-wrapper.sh` runs with `LOOM_MAX_RETRIES=1` (set by the shepherd to avoid retry multiplication — see #2516), and the startup monitor kills the session due to an MCP failure, the wrapper:

1. Correctly detects `"transient error pattern: MCP server failed"`
2. Has no retry budget left (`max_retries=1`)
3. Exits with **code 1** (generic failure)

The Python `run_phase_with_retry()` only retries specific exit codes:
- 4 = stuck
- 6 = low output  
- 7 = MCP failure
- 10 = ghost session

Exit code 1 falls through as a generic error — the builder phase fails without retry at the Python level, even though the failure was a transient MCP issue that would likely succeed on retry.

## Observed in

Shepherd run for #2729: builder subprocess was killed by startup monitor due to global plugin MCP failure (pre-#2726 code in the worktree). The wrapper logged `[INFO] Detected transient error pattern: MCP server failed` then `[ERROR] Max retries (1) exceeded` and exited 1.

## Why it matters less now

PR #2726 tolerates global plugin failures when project MCPs connect, so the specific trigger (startup monitor killing session for non-project MCP failures) should no longer occur. This is defense-in-depth.

## Proposed fix

When `claude-wrapper.sh` exhausts retries and the last failure was classified as an MCP transient error, exit with code **7** instead of code **1**. This allows the Python retry layer to recognize and retry the MCP failure properly.

Alternatively, have the Python `run_phase_with_retry()` inspect the wrapper log for MCP failure markers when exit code is 1, similar to how it already classifies ghost session causes.

--- a/defaults/scripts/claude-wrapper.sh
+++ b/defaults/scripts/claude-wrapper.sh
@@ -712,6 +712,26 @@ is_transient_error() {
     return 1
 }
 
+# Check if error output specifically indicates an MCP/plugin failure.
+# Used to map exhausted-retry exits to exit code 7 so the Python retry
+# layer can recognize MCP failures even when the wrapper's own retries
+# are exhausted.  See issue #2746.
+is_mcp_error() {
+    local output="$1"
+    local mcp_patterns=(
+        "MCP server failed"
+        "MCP.*failed"
+        "plugins failed"
+        "plugin.*failed to install"
+    )
+    for pattern in "${mcp_patterns[@]}"; do
+        if echo "${output}" | grep -qi "${pattern}"; then
+            return 0
+        fi
+    done
+    return 1
+}
+
 # Monitor output file for API errors during execution.
 # If an API error pattern is detected and no new output arrives within
 # API_ERROR_IDLE_TIMEOUT seconds, sends SIGINT to the claude process.
@@ -1182,6 +1202,15 @@ run_with_retry() {
     log_error "Max retries (${MAX_RETRIES}) exceeded"
     log_error "Last error: ${output}"
     clear_retry_state
+
+    # When the last failure was MCP-related, exit with code 7 so the
+    # Python retry layer (run_phase_with_retry) can recognise and retry
+    # the MCP failure instead of treating it as a generic error.
+    # See issue #2746.
+    if is_mcp_error "${output}"; then
+        log_info "Last failure was MCP-related — exiting with code 7"
+        return 7
+    fi
     return 1
 }
 

